### PR TITLE
Add an OpSpec lab: capture, run and explain OpSpec lists standalone

### PR DESCRIPTION
**Pull request by willmj**
#### What type of PR is this?

- [ ] bug
- [x] feature
- [x] documentation
- [ ] cleanup

#### What this PR does:

Adds a standalone lab for inspecting and re-running OpSpec lists.

Three tools under `tests/op_specs/`:

| Tool | Job |
|---|---|
| `capture.py` | Point it at any PyTorch script; get one self-contained, runnable file per compiled kernel |
| `runner.py` | The run loop from OpSpec list to device launch, with no `torch.compile` involved |
| `explain.py` | Render a resolved OpSpec as a readable summary instead of a wall of symbolic expressions |

Plus four ~15-line programs in `tests/op_specs/programs/` (elementwise add, fused softmax, matmul, and a matmul-softmax-matmul chain that spills to the HBM pool), and a new user-guide page that walks them.

The point is bisection. Running an OpSpec in isolation splits a wrong result in two: if the captured spec is *also* wrong, the back end is mis-executing a faithful description of the intent, and you have a minimal reproducer with no model, Dynamo or Inductor left in it. If it comes out correct, the bug is upstream in lowering, fusion, stickification or layout. This plugs into Step 4 of the existing debugging workflow - `sendnn` tells you which side, a captured OpSpec gives you a reproducer of that side.

#### Which issue(s) this PR is related to:

Issue #3585 

#### Special notes for your reviewer:

Suggested review order, which is also dependency order: `runner.py` → `capture.py` → `explain.py` → the four programs → the guide.

Two facts about the IR explain why `explain.py` exists, and they are the two things worth understanding before reading it:

- A **layout label is an equivalence class over layouts, not an argument role.** One label per distinct layout, so `raw SDSC label "OUTPUT"` routinely appears on *inputs*. And `LAYOUT_LABELS` and `MATMUL_LAYOUT_LABELS` disagree on order, so the same index yields a different string depending on the op. The view relabels to `L0`/`L1`/… and discloses the raw name in a footnote.
- **Iteration symbols are renamed positionally.** `parse_op_spec` zips `iteration_space` insertion order against `_get_op_dim_labels()`, so the dict's order is load-bearing while the symbol names are not - `c0`/`c1` in one kernel, `d0`/`d1` in another, both becoming `mb`/`out`.

On the tests: `test_op_spec_lab.py` is deliberately a tripwire rather than a specification, and the device paths in `runner.py` have no unit coverage.

Captures are git-ignored rather than committed.

**Verification.** Unit tests and lint on a Spyre pod:

```
python -m pytest tests/op_specs/test_op_spec_lab.py -q
pre-commit run --files .gitignore pyproject.toml \
    docs/source/user_guide/debugging/op_spec_lab.md \
    tests/op_specs/test_op_spec_lab.py tests/op_specs/programs/*.py
```

7 passed; `ruff check`, `ruff format`, `PyMarkdown` and `mypy` all pass with no file modified by a hook.

End-to-end on device, which is what verified the guide:

```
for p in 01_add 02_softmax 03_matmul 04_pool_chain; do
    python tests/op_specs/programs/$p.py
    python tests/op_specs/capture.py tests/op_specs/programs/$p.py
done
python captured/sdsc_fused_add_0.py --stage bundle --out-dir /tmp/bundle01
```

All four programs and captures succeed. Kernel names, OpSpec and argument counts, the 32768-byte pool and every quoted explain block match the guide. `--stage bundle` writes `sdsc_0.json` and `bundle.mlir`.

Every transcript in the guide is real device output rather than an illustration; the programs pin seed, shapes and dtypes so the numbers reproduce. Each "change one thing" step was executed rather than reasoned about, which is how one got corrected: shrinking 04's chain scales the pool from 32768 to 8192 bytes rather than eliminating it - the opposite of what the earlier draft implied. The others hold: 01 at 320 columns gives `ceil(320/64) = 5` and `device_size [5, 128, 64]`, and 02 at `(32, 256)` gives a reduction output of `[1, 32, 64]`.

#### Does this PR introduce a user-facing change?

Yes, but additive only and with no behavioural change.

- **New user-guide page:** "The OpSpec Lab" under `docs/source/user_guide/debugging/`, linked from the debugging index.
- **New developer tooling:** `capture.py`, `runner.py` and `explain.py` under `tests/op_specs/`, invoked directly by hand - not imported by the library and not part of any public API.
- **Removed:** `tests/op_specs/op_spec.py`, a hand-written sample spec superseded by `capture.py`.
- No change to any `torch_spyre` API, op behaviour or numerics.

#### Additional note:

Authored with the assistance of an AI coding agent. All changes reviewed by the committer before submission.

### CODE REVIEW

**Review by dgrove-oss — DISMISSED** [review]
Summary

Adds a self-contained debugging toolkit under tests/op_specs/: capture.py turns any PyTorch script into standalone, runnable OpSpec scripts (one per compiled kernel); runner.py provides the run loop (compile → device launch) with no torch.compile involved, and is inlined verbatim into every captured script; explain.py renders a resolved OpSpec as a readable summary. Four small example programs plus a new user-guide page (docs/source/user_guide/debugging/op_spec_lab.md) walk through the tool end to end. A small refactor in torch_spyre/_inductor/spyre_kernel.py extracts uses_hbm_pool() as a shared, importable predicate. The stale hand-written tests/op_specs/op_spec.py sample is removed, superseded by capture.py.

Review

BLOCKERs

None.

SUGGESTIONs

- The pool-contents limitation (pool.cpu() aborts the process, so captured pool contents are always uninitialized on replay) is documented in three separate docstrings/comments (capture.py's _record_args, runner.py's make_pool, the guide). Consider consolidating into one canonical explanation referenced from the others, so a future fix to that limitation doesn't require updating three prose copies in sync.
- explain.py's column-budget helpers (_labelled, _wrapped, _fit, _table) are intricate enough (as test_hostile_names_stay_within_width's regression list attests) that a short module-level note pointing at that test as the "if you touch wrapping logic, re-run this" tripwire would help future editors — right now that link is implicit.
- explain.py's _sticks() deliberately lets elems_per_stick() raise uncaught (documented rationale: silent disappearance is worse than a visible !! line), which is a good call, but it's the one section doing this while its siblings return [] on a miss — worth a one-line comment at the call site in _op_block confirming this asymmetry is intentional, for whoever adds the fifth section.

QUESTIONs

- capture.py's dedup logic strips debug_handle lines before comparing specs (to collapse an eager-vs-compiled double-capture of the same kernel). Is there a known case where two genuinely different kernels could collide under this key (same rendered spec modulo provenance, different intent)? If not, no action needed — just confirming the invariant is as tight as it looks.
- The PR body notes runner.py's device paths have no unit coverage (by design, since CI here is expected to run without a Spyre device for the docs/lint/build-only checks, but a device is available in this repo's CI per gh pr checks). Is there a follow-up planned to exercise run_op_specs/bundle_op_specs on the hardware-backed CI lane, or is that intentionally deferred to manual verification only (as done in this PR's description)?

VERDICT

Approve

**Comment by willmj** [comment]
> Is there a known case where two genuinely different kernels could collide under this key (same rendered spec modulo provenance, different intent)?

No, `_codegen_op_spec_list` emits every field codegen consumes (`op`, `is_reduction`, `iteration_space`, `op_info`, `tiled_symbols`, `tiled_symbol_trip_counts`, `node_output_ranges`, `args`) and the key also includes the kernel name, so `debug_handle` is the only thing stripped and it never reaches the backend.

> Is there a follow-up planned to exercise run_op_specs/bundle_op_specs on the hardware-backed CI lane, or is that intentionally deferred to manual verification only (as done in this PR's description)?

Intentionally deferred to manual verification for now - happy to open a follow-up issue to run `run_op_specs`/`bundle_op_specs` on the hardware-backed lane if that's wanted.

**Review by dgrove-oss — DISMISSED** [review]
lgtm

**Review by dgrove-oss — DISMISSED** [review]
Summary

Adds a self-contained debugging toolkit (tests/op_specs/{capture,runner,explain}.py) that lets a developer capture the OpSpec list produced for a failing kernel as a standalone script, replay just the backend half of compilation, and render the specs as human-readable text — isolating whether a bug lives in the backend/codegen or further upstream in lowering/fusion/stickification. Includes a 566-line user guide, four worked example programs, a comprehensive pure-Python test suite, and a small, behavior-preserving refactor of spyre_kernel.py to share the HBM-pool-detection predicate with the new tooling.

Review

License headers, DCO, and pre-commit compliance all check out (see below). The design is careful about a real correctness hazard: captured scripts inline their run-loop helpers via inspect.getsource() (runner.runner_template()) rather than duplicating source as a string literal, so a future edit to runner.py can't silently make old captures stale — and test_runner_template_is_self_contained walks the AST to prove every name the inlined loop uses actually resolves in a bare script, not just that it compiles. The explain.py width-budget test (test_hostile_names_stay_within_width) reconstructs three specific historical overflow regressions as concrete inputs rather than asserting against a canned number, and correctly treats an unbreakable long identifier as an exempted case ("printed whole, never truncated") rather than silently truncating it. capture.py's handling of the HBM-pool tensor (never calling .cpu() on it, since that corrupts the heap) is documented inline and shares the same uses_hbm_pool() predicate now used by spyre_kernel.py's call_kernel, so the two can't drift apart on what "uses the pool" means.

BLOCKERs

- None found.

SUGGESTIONs

- None — this is thorough, well-tested, self-contained tooling with no scope creep into unrelated files.

QUESTIONs

- None — the PR body's own review-order guidance (runner → capture → explain → programs/guide) matched what a fresh read of the diff needed, and following it answered everything I'd otherwise have asked.
                                                                                     
Verdict
APPROVE

**Comment by mudhakar** [comment]
1. build_tensors() can't synthesize fp8 args (capture.py:426): torch.rand has no uniform kernel for them
2. Beat-4 transcripts assume a clean captured/. should it be removed every time or use --out?
3. capture.main (512-18, 581-87) returns 0 after the target crashed - use a distinct exit code?

**Review by mudhakar — APPROVED** [review]
LGTM

**Review by dgrove-oss — APPROVED** [review]
lgtm
